File: src/lib/specialists.ts
/**
 * v3 verification pipeline: parallel specialist agents + synthesis.
 * Each specialist receives Tavily web excerpts in the user message; synthesis merges memos.
 */

export const SPECIALIST_KEYS = ["fundamentals", "news", "macro", "sentiment"] as const;
export type SpecialistKey = (typeof SPECIALIST_KEYS)[number];

const SPECIALIST_CORE = `
You are an NGX (Nigerian Exchange) research specialist. Output Markdown only — no preamble or meta-commentary.

Rules:
- Use only the web search excerpts in the user message below for factual claims (URLs/snippets). Prefer ngxgroup.com, Nairametrics, Proshare Nigeria, BusinessDay Nigeria, CBN when present in excerpts.
- The ticker is an **NGX (Nigerian Exchange)** listing. If a snippet clearly refers to a different country or exchange (e.g. US/EU homonym symbol), ignore it unless it explicitly ties to Nigeria or NGX.
- Never fabricate figures. If data is missing, say "Data not found" for that item.
- Prefer sources from the last 90 days; note if older.
- At the very end of your output, on its own line, include exactly one confidence marker:
  [High] — reason
  [Medium] — reason
  [Low] — reason
  (Same definitions as TickerNG: High = 2+ independent dated sources; Medium = one source or stale; Low = inferred or none.)
`.trim();

export function getSpecialistSystemPrompt(key: SpecialistKey): string {
  const briefs: Record<SpecialistKey, string> = {
    fundamentals: `
${SPECIALIST_CORE}

Your scope ONLY: company fundamentals and valuation for the given ticker.
Cover in Markdown with short ### subsections:
- Company snapshot: sector, market cap if found, average daily volume, float/liquidity note. Flag if volume is below 10,000 shares/day or unknown.
- Financial summary: revenue trend (2–3 years if available), EPS, margins. Note conflicts between sources.
- Valuation: P/E, P/B, vs peers or history if found.

Do NOT cover macro (CBN, inflation), news headlines, or analyst ratings here — another agent handles those.
`.trim(),

    news: `
${SPECIALIST_CORE}

Your scope ONLY: material news and corporate actions for the given NGX ticker.
Cover in Markdown:
- Notable news in the last ~90 days (results, guidance, regulatory, M&A, leadership).
- Corporate actions: dividends, AGMs, splits/bonuses if any.
- Upcoming dates (results, AGM) if found.

Do NOT duplicate full financial tables or valuation — reference briefly if needed for context only.
`.trim(),

    macro: `
${SPECIALIST_CORE}

Your scope ONLY: macro and market context relevant to this company's sector on the NGX.
Cover in Markdown:
- CBN Monetary Policy Rate (current if found)
- Nigeria headline inflation (latest release)
- FX / naira context if relevant to the sector
- NGX All-Share Index trend (recent / YTD if found)
- One short paragraph on how macro might affect this sector or ticker (qualitative, no stock price predictions)

Do NOT write full company financials or analyst ratings.
`.trim(),

    sentiment: `
${SPECIALIST_CORE}

Your scope ONLY: analyst and market sentiment for the given ticker.
Cover in Markdown:
- Any Buy/Hold/Sell or equivalent ratings with source and date. If none: state "No recent analyst consensus found" — do not infer.
- Brief note on retail/social sentiment ONLY if you find a reliable, citable source; otherwise say "No reliable social sentiment source found."

Do not repeat full fundamentals; reference the investment case in one or two lines max if needed.
`.trim(),
  };
  return briefs[key];
}

export function getSpecialistUserMessage(ticker: string, key: SpecialistKey): string {
  const t = ticker.toUpperCase();
  const labels: Record<SpecialistKey, string> = {
    fundamentals: "Fundamentals & valuation",
    news: "News & corporate actions",
    macro: "Macro & market context",
    sentiment: "Analyst & sentiment",
  };
  return `Ticker: ${t}\n\nProduce the ${labels[key]} specialist memo for this NGX stock using the excerpts below. End with a single [High|Medium|Low] — reason line.`;
}

const SYNTHESIS_RULES = `
You are the lead editor for TickerNG verification reports (Nigerian Exchange). You receive four specialist memos (fundamentals, news, macro, sentiment) produced by parallel researchers. You do NOT have web search — work only from the memos.

Output a single Markdown verification report for the user. No preamble.

Rules:
- Merge overlapping facts; where specialists conflict, present both and note the discrepancy (do not pick one silently).
- If a specialist section failed or is empty, write a short note in the relevant part of the report (e.g. "Limited specialist data for news") and use [Low] confidence for that slice.
- Use the EXACT section headers below, in order.
- Each of sections 1–5 must end with a confidence line on its own line:
  [High] — reason
  [Medium] — reason
  [Low] — reason
- Section 6 (Verdict) must also end with one confidence line.
- Follow TickerNG style: tables where helpful, concise prose, no chat filler.

## Required report structure

### 1. Company Snapshot
(Draw primarily from fundamentals memo; align with news if relevant.)
- [Confidence — reason]

### 2. Financial Summary
(From fundamentals memo.)
- [Confidence — reason]

### 3. Valuation
(From fundamentals memo.)
- [Confidence — reason]

### 4. Catalysts
(Primarily from news memo; tie to fundamentals if needed.)
- [Confidence — reason]

### 5. Analyst Sentiment
(From sentiment memo.)
- [Confidence — reason]

### 6. Verdict
- 🟢 Bull case: 2–3 concrete points grounded in the memos.
- 🔴 Bear case: 2–3 concrete points.
- 🚩 Red flags: only if supported; omit subsection entirely if none — do not write "None identified."
- 💧 Liquidity note: from fundamentals memo (volume vs 10,000 shares/day threshold).
- [Confidence — reason]

Respond with only the report in Markdown.
`.trim();

export function getSynthesisSystemPrompt(): string {
  return SYNTHESIS_RULES;
}

export function getSynthesisUserMessage(ticker: string, memos: Record<SpecialistKey, string>): string {
  const t = ticker.toUpperCase();
  return `Produce the final verification report for NGX ticker ${t}.

<fundamentals_memo>
${memos.fundamentals}
</fundamentals_memo>

<news_memo>
${memos.news}
</news_memo>

<macro_memo>
${memos.macro}
</macro_memo>

<sentiment_memo>
${memos.sentiment}
</sentiment_memo>

Integrate macro context naturally into sections 1–5 where relevant (especially Company Snapshot and Catalysts); you may add a short macro sentence under Company Snapshot or Catalysts rather than a duplicate standalone macro section — the user-facing structure must remain as specified above with those six ### sections only.`;
}

File: api/agent/stages/specialists.py
"""Per-specialist Tavily context + prompt fragments for verification."""

from concurrent.futures import ThreadPoolExecutor

from agent.config.constants import GLOBAL_TICKER_NOISE_DOMAINS, NGX_EXCHANGE_DOMAINS, NGX_PRIORITY_DOMAINS
from agent.prompts.prompt_resolve import specialist_core
from agent.search.tavily import tavily_depth, tavily_search_to_markdown
from agent.stages.queries import (
    filings_exchange_query,
    macro_query,
    market_data_query,
    narrow_query,
    sentiment_deep_query,
    wide_query,
)


def fundamentals_tavily_markdown(api_key: str, ticker: str) -> str:
    noise = GLOBAL_TICKER_NOISE_DOMAINS
    futures = []
    with ThreadPoolExecutor(max_workers=4) as ex:
        futures.append(
            ex.submit(
                tavily_search_to_markdown,
                api_key,
                narrow_query(ticker, "fundamentals"),
                {
                    "topic": "finance",
                    "maxResults": 12,
                    "searchDepth": tavily_depth("default"),
                    "includeDomains": NGX_PRIORITY_DOMAINS,
                    "sectionHeading": "Web search (A): NGX & Nigerian financial / news sources",
                },
            )
        )
        futures.append(
            ex.submit(
                tavily_search_to_markdown,
                api_key,
                wide_query(ticker, "fundamentals"),
                {
                    "topic": "finance",
                    "country": "nigeria",
                    "maxResults": 12,
                    "searchDepth": tavily_depth("default"),
                    "excludeDomains": noise,
                    "sectionHeading": "Web search (B): Nigeria-wide (homonym-safe query)",
                },
            )
        )
        futures.append(
            ex.submit(
                tavily_search_to_markdown,
                api_key,
                filings_exchange_query(ticker),
                {
                    "topic": "finance",
                    "maxResults": 15,
                    "searchDepth": tavily_depth("filing"),
                    "includeDomains": NGX_EXCHANGE_DOMAINS,
                    "sectionHeading": "Web search (C): NGX & FMDQ - filings, reports, key figures",
                },
            )
        )
        futures.append(
            ex.submit(
                tavily_search_to_markdown,
                api_key,
                market_data_query(ticker),
                {
                    "topic": "finance",
                    "maxResults": 10,
                    "searchDepth": tavily_depth("market"),
                    "country": "nigeria",
                    "excludeDomains": noise,
                    "sectionHeading": "Web search (D): market data - price, cap, volume (NGX context)",
                },
            )
        )
        return "\n".join(f.result() for f in futures)


def build_specialist_web_context(api_key: str, ticker: str, key: str) -> str:
    noise = GLOBAL_TICKER_NOISE_DOMAINS
    t = ticker.upper()
    if key == "macro":
        return tavily_search_to_markdown(
            api_key,
            macro_query(ticker),
            {
                "topic": "general",
                "country": "nigeria",
                "maxResults": 10,
                "searchDepth": "basic",
                "excludeDomains": noise,
                "sectionHeading": "Web search: macro & Nigerian market context",
            },
        )
    if key == "fundamentals":
        return fundamentals_tavily_markdown(api_key, ticker)
    if key == "news":
        with ThreadPoolExecutor(max_workers=3) as ex:
            a = ex.submit(
                tavily_search_to_markdown,
                api_key,
                narrow_query(ticker, key),
                {
                    "topic": "general",
                    "maxResults": 8,
                    "searchDepth": tavily_depth("default"),
                    "includeDomains": NGX_PRIORITY_DOMAINS,
                    "sectionHeading": "Web search (A): NGX & Nigerian financial / news sources",
                },
            )
            b = ex.submit(
                tavily_search_to_markdown,
                api_key,
                wide_query(ticker, key),
                {
                    "topic": "general",
                    "country": "nigeria",
                    "maxResults": 10,
                    "searchDepth": tavily_depth("default"),
                    "excludeDomains": noise,
                    "sectionHeading": "Web search (B): Nigeria-wide (homonym-safe query)",
                },
            )
            c = ex.submit(
                tavily_search_to_markdown,
                api_key,
                f"{t} Nigeria NGX listed company stock news",
                {
                    "topic": "news",
                    "maxResults": 7,
                    "searchDepth": tavily_depth("default"),
                    "timeRange": "month",
                    "excludeDomains": noise,
                    "sectionHeading": "Web search (C): news topic (recent)",
                },
            )
            return f"{a.result()}\n{b.result()}\n{c.result()}"
    if key == "sentiment":
        with ThreadPoolExecutor(max_workers=3) as ex:
            a = ex.submit(
                tavily_search_to_markdown,
                api_key,
                narrow_query(ticker, key),
                {
                    "topic": "general",
                    "maxResults": 8,
                    "searchDepth": tavily_depth("default"),
                    "includeDomains": NGX_PRIORITY_DOMAINS,
                    "sectionHeading": "Web search (A): NGX & Nigerian financial / news sources",
                },
            )
            b = ex.submit(
                tavily_search_to_markdown,
                api_key,
                wide_query(ticker, key),
                {
                    "topic": "general",
                    "country": "nigeria",
                    "maxResults": 10,
                    "searchDepth": tavily_depth("default"),
                    "excludeDomains": noise,
                    "sectionHeading": "Web search (B): Nigeria-wide (homonym-safe query)",
                },
            )
            c = ex.submit(
                tavily_search_to_markdown,
                api_key,
                sentiment_deep_query(ticker),
                {
                    "topic": "finance",
                    "maxResults": 10,
                    "searchDepth": tavily_depth("sentimentDeep"),
                    "country": "nigeria",
                    "excludeDomains": noise,
                    "sectionHeading": "Web search (C): analyst, coverage, and rating language",
                },
            )
            return f"{a.result()}\n{b.result()}\n{c.result()}"
    raise RuntimeError(f"Unhandled specialist: {key}")


def get_specialist_system_prompt(key: str) -> str:
    core = specialist_core()
    if key == "fundamentals":
        return f"""{core}

Your scope ONLY: company fundamentals and valuation for the given ticker.
Cover: snapshot (sector/mcap/volume), financial summary, valuation.
Do NOT cover macro/news/sentiment in depth.
"""
    if key == "news":
        return f"""{core}

Your scope ONLY: material news and corporate actions for the given ticker.
Cover notable news, corporate actions, upcoming dates.
"""
    if key == "macro":
        return f"""{core}

Your scope ONLY: macro and market context relevant to this ticker/sector.
Cover CBN rate, inflation, FX context, NGX All-Share trend, and one qualitative impact paragraph.
"""
    return f"""{core}

Your scope ONLY: analyst and market sentiment for the given ticker.
Cover ratings/consensus if found, and reliable social sentiment if available.
"""


def get_specialist_user_message(ticker: str, key: str) -> str:
    labels = {
        "fundamentals": "Fundamentals & valuation",
        "news": "News & corporate actions",
        "macro": "Macro & market context",
        "sentiment": "Analyst & sentiment",
    }
    return f"Ticker: {ticker.upper()}\n\nProduce the {labels[key]} specialist memo for this NGX stock using the excerpts below. End with a single [High|Medium|Low] - reason line."
